### ISSUE TITLE
parallel netcdf writes

### ISSUE THREAD

**Issue by jswhit2** [NOAA-EMC/ufsatm#23]
Parallel write capability is needed in module_write_netcdf.F90.  The current version of the netcdf library does not support parallel writing of compressed files, but uncompressed parallel writes should work.  Here are some steps needed to implement this:

1) add a flag to model_configure to indicate parallel IO is desired (for now make sure this flag is set to false if compression is enabled).
2) if parallel IO is enabled, open the file using nf90_create on all tasks and pass the optional mpi_comm and mpi_info arguments.
3) the nf90_put_var calls need to be modified to write independent slices (defined by istart,jstart,iend,jend,kstart,kend).  The ESMF_Gather call should be skipped.

**Comment by junwang-noaa** [comment]
Jeff,

Just a quick update, I am working on this following the steps you listed here. The coding is mostly done. One thing not listed in your steps and not clear to me  is about writing attribute, I assume it is OK to call nf90_put_att on all the mpi tasks?

**Comment by jswhit** [comment]
Yes, creating the attributes, dimensions and variables can be done  on all tasks.  The only thing that needs to change is the nf90_put_var (writing data to the variable).

**Comment by junwang-noaa** [comment]
Great! Currently I am going to test the code on dell as it has the parallel
netcdf lib Cory built. Will copy the code to hera for you to check when the
coding is done. Thanks.

**Comment by jswhit** [comment]
Jun - one thing to be aware of.  There are two parallel IO implementations with the same API - one that works with classic (non-hdf5) files and one that works with hdf5 files.  To enable both, you have to build with `--enable-parallel4` *and*  `--enable-pnetcdf`.  If Cory did not build with `--enable-pnetcdf` you will have to make this change

```fortran
-    if (ideflate == 0) then
-        ncerr = nf90_create(trim(filename), &
-        cmode=IOR(IOR(NF90_CLOBBER,NF90_64BIT_OFFSET),NF90_SHARE), &
-        ncid=ncid); NC_ERR_STOP(ncerr)
-        ncerr = nf90_set_fill(ncid, NF90_NOFILL, oldMode); NC_ERR_STOP(ncerr)
-    else
-        ncerr = nf90_create(trim(filename), cmode=IOR(IOR(NF90_CLOBBER,NF90_NETCDF4),NF90_CLASSIC_MODEL), &
-        ncid=ncid); NC_ERR_STOP(ncerr)
-        ncerr = nf90_set_fill(ncid, NF90_NOFILL, oldMode); NC_ERR_STOP(ncerr)
-    endif
+   ncerr = nf90_create(trim(filename), cmode=IOR(IOR(NF90_CLOBBER,NF90_NETCDF4),NF90_CLASSIC_MODEL), &
+   ncid=ncid); NC_ERR_STOP(ncerr) ! modify if parallel IO needed
+   ncerr = nf90_set_fill(ncid, NF90_NOFILL, oldMode); NC_ERR_STOP(ncerr)
```

**Comment by junwang-noaa** [comment]
I see, thanks for letting me know.

Jun

**Comment by climbfuji** [comment]
For your information, the pnetcdf (parallel-netcdf) library developed at Argonne Labs writes data in netCDF cdf-2 format (called 64-bit offset in pnetcdf; is essentially the same as netCDF 3 format and can be read and written with any unidata netCDF library from 3.x upwards, but parallel reads/writes require pnetcdf support as described by Jeff for netCDF 4.x). Note that early versions of netcdf 4 (4.5.x) had some serious bugs when writing to those cdf-5 files, which in some cases corrupted the data.

For larger files (individual variables larger than 2GB), the pnetCDF format must be switched to 64-bit data format (aka cdf-5; NF90_64BIT_DATA instead of NF90_64BIT_OFFSET in the code snippet above), which is incompatible with standard netCDF installations and also requires building newer unidata netcdf4.x versions with the flags that Jeff described above. The drawback of this format is that many other downstream tools don't recognize it (you will see an error message of the kind "MPI routine called before MPI init" or so) - unless they were built using the same pnetcdf-enabled version of netCDF 4.x.

The other format using parallel hdf5 as backend for so-called netCDF4 or netCDF4-classic files. This does't require the parallel-netcdf backend and seems to be easier. BUT: the parallel reading and writing of netCDF4 files through the HDF5 format is several times slower than that through parallel-netCDF!

Long story short: for the given "small" global meshes of 13km resolution, we can use the CDF-2  parallel-netcdf version, which can be read or written with any existing software that can read netCDF3 classic files. We should not use the netCDF4-phdf5 backend, because it is very, very slow. I have some HPC reports (unpublished work) from my previous job where I was using pnetcdf, netcdf4-phdf5 and SIONlib in MPAS extreme scaling experiments if you are interested.

**Comment by jswhit** [comment]
We have to use the hdf5 backend to get compression.  My experience is that parallel-hdf5 performance has improved quite a bit in recent versions.

**Comment by climbfuji** [comment]
I forgot about the compression part. Good to hear that there has been a speedup in recent versions. The real test will be when we scale up the model to 3-5km global resolution and 100k +/- MPI tasks.

Out of curiosity, have you tried different compression levels (1-9 if I remember correctly) in terms of smaller file sizes versus loss in read/write performance?

Thanks for implementing this parallel read/write capability, dearly needed.

**Comment by jswhit** [comment]
We're using a compression level of 1.

**Comment by edwardhartnett** [comment]
Howdy!

To clarify a few points: 
* The CDF5 format originally developed by Argonne for pnetcdf is now part of the Unidata netCDF library as well, and is a canonical and supported binary format of netCDF. Older versions of netCDF will not understand it, but all recent and future versions do. So it is safe to use and distribute, however offers no compression.

* I have a PR open on the netcdf-c projec that will allow parallel writes with zlib, and that should help.

* I am going to re-instate the szip filter in the netCDF C library as well. This will allow szip to be easily used. Since szip was supported in netCDF in read-only mode, all existing netCDF installs will be able to read these files, as long as HDF5 was installed with szip capability.

* I am exploring some new filters, specifically LZ4. I should offer much better read and write performance, at a cost of slightly less good compression. The challenge is making it available to everyone, but this is something we are working out.

* HDF5 parallel I/O should be reasonably fast, when settings are correct. If you build netcdf-c with the --enable-benchmarks option, you get a program nc_perf/bm_file, which will allow you to test your file read and write time with a variety of chunksize settings, so you can get a feel for how it changes performance. Let me know if you want help.

**Comment by climbfuji** [comment]
> Howdy!
> 
> To clarify a few points:
> 
> * The CDF5 format originally developed by Argonne for pnetcdf is now part of the Unidata netCDF library as well, and is a canonical and supported binary format of netCDF. Older versions of netCDF will not understand it, but all recent and future versions do. So it is safe to use and distribute, however offers no compression.
> * I have a PR open on the netcdf-c projec that will allow parallel writes with zlib, and that should help.
> * I am going to re-instate the szip filter in the netCDF C library as well. This will allow szip to be easily used. Since szip was supported in netCDF in read-only mode, all existing netCDF installs will be able to read these files, as long as HDF5 was installed with szip capability.
> * I am exploring some new filters, specifically LZ4. I should offer much better read and write performance, at a cost of slightly less good compression. The challenge is making it available to everyone, but this is something we are working out.
> * HDF5 parallel I/O should be reasonably fast, when settings are correct. If you build netcdf-c with the --enable-benchmarks option, you get a program nc_perf/bm_file, which will allow you to test your file read and write time with a variety of chunksize settings, so you can get a feel for how it changes performance. Let me know if you want help.

Thanks for the update, Ed. Did you ever get to play with the SIONlib backend and test its performance? We had a telecon with the developers in Juelich some time ago, and I didn't have any time to follow up on this, sorry.

**Comment by edwardhartnett** [comment]
No I have not played with SIONlib but would be interested in learning more about it. There was an idea of writing a SIONlib read/write module for netcdf...

**Comment by edwardhartnett** [comment]
OK, some more news: turns out that szip is already enabled in netcdf-c for writes! It's a bit of an undocumented feature.

For this to work, HDF5 must be built with szip (as well as the usual zlib).

Once netcdf-c is built on a HDF5 that supports szip, then szip compression may be turned on for a var like this:
```
#define HDF5_FILTER_SZIP 4
    /*
     * Set parameters for SZIP compression; check the description of
     * the H5Pset_szip function in the HDF5 Reference Manual for more
     * information.
     */
    szip_params[0] = H5_SZIP_NN_OPTION_MASK;
    szip_params[1] = H5_SZIP_MAX_PIXELS_PER_BLOCK_IN;
    stat = nc_def_var_filter(ncid, varid, HDF5_FILTER_SZIP, 2, szip_params);

```

Currently this will only work for sequential access. I am working on getting it working for parallel access next. ;-)

**Comment by junwang-noaa** [comment]
That is great news! Thanks for letting us know!

Jun

**Comment by jswhit2** [comment]
@junwang-noaa - do you have a branch with the parallel-io mods in it that I can play with?

**Comment by junwang-noaa** [comment]
Yes. The code gets compiled on mars, but I haven't tested the parallel netcdf case yet. 

[submodule "FV3"]
        path = FV3
        url = https://github.com/junwang-noaa/fv3atm
        branch = netcdf_parallel

The branch also has the code changes for real(8) lon/lat in netcdf file and a bug fix for post.

**Comment by jswhit2** [comment]
Thanks @junwang-noaa, I will give it a spin on hera.

**Comment by jswhit2** [comment]
I've made a few changes to @junwang-noaa's netcdf_parallel branch and *almost* have it running on hera (without compression).  The files are created, but the data isn't actually written.  Jun - if you give me access to your fork I can push my changes there, or if you prefer I can create my own fork.

**Comment by junwang-noaa** [comment]
I added you as a collaborator, once you accept the invitation, you should
be able to push to the branch. Please let me know if you still can't.

Jun

**Comment by jswhit2** [comment]
I've updated https://github.com/junwang-noaa/fv3atm/tree/netcdf_parallel, to include bug fixes and changes from PR #18.  It now runs on hera, and for uncompressed data shows some significant speedups.  Next step is to build the netcdf library with https://github.com/Unidata/netcdf-c/pull/1582 so we can test parallel compressed writes.

**Comment by jswhit2** [comment]
BTW - the code is now using independent (not collective) parallel access.  Collective access seems to be quite a bit slower in my tests - don't understand why.  Using independent access required changing the time dimension from unlimited to fixed length.

**Comment by edwardhartnett** [comment]
Collective access is required for any filters in HDF5, including all compression filters. :-(

**Comment by jswhit2** [comment]
OK - I've updated the code to turn collective access for variables that are compressed.

**Comment by junwang-noaa** [comment]
Ed,

Would we have a netcdf beta library that we can test in our FV3 parallel?
Thanks

Jun

**Comment by jswhit2** [comment]
Jun - it's not merged into master yet but we can check out https://github.com/NOAA-GSD/netcdf-c/tree/ejh_parallel_zlib.  I'll build this on hera.

**Comment by junwang-noaa** [comment]
Jeff,

Do you have instructions on how to build it? I can build it on wcoss.
Thanks.

Jun

**Comment by jswhit2** [comment]
Jun: Here's what I did on hera (haven't yet tested it)

* create a directory ${parlibpath}
* cd ${parlibpath}
* download hdf5-1.10.6.tar.gz and netcdf-fortran-4.5.2.tar.gz to that directory

To build HDF5:

* tar -xvzf hdf5-1.10.6.tar.gz
* cd hdf5-1.10.6
* ./configure --prefix=${parlibpath} --enable-hl --enable-parallel
* make
* make install

To build netcdf-c:

* git clone https://github.com/NOAA-GSD/netcdf-c
* cd netcdf-c; git checkout ejh_parallel_zlib
* autoreconf -i
* setenv LDFLAGS -L${parlibpath}/lib
* setenv CPPFLAGS -I${parlibpath}/include
* ./configure --prefix=${parlibpath} --enable-netcdf-4 --enable-shared --disable-dap --enable-parallel4
* make
* make install

To build netcdf-fortran:

* tar -xvzf netcdf-fortran-4.5.2.tar.gz
* cd netcdf-fortran-4.5.2
* setenv FC mpif90
* setenv CC mpicc
* setenv LDFLAGS -L${parlibpath}/lib
* setenv CPPFLAGS -I${parlibpath}/include
* ./configure --prefix=${parlibpath}
* make
* make install

**Comment by climbfuji** [comment]
Just for your information,

I compiled the entire parallel I/O stack on hera for some other folks using MPAS and for my own 3km-global MPAS tests. I made those available as modules:

module use -a /scratch1/BMC/gmtb/software/modulefiles/intel-18.0.5.274/impi-2018.0.4

module load intel/18.0.5.274
module load impi/2018.0.4
module load pnetcdf/1.11.2-for-pio
module load netcdf/4.7.0-for-pio
# module load pio/2.4.4 # not needed for FV3
module list

Maybe that is good enough for your testing?

Thanks,

Dom

**Comment by jswhit2** [comment]
Dom: We need a bleeding edge version of netcdf-c from  https://github.com/NOAA-GSD/netcdf-c/tree/ejh_parallel_zlib compiled with parallel hdf support.

**Comment by jswhit2** [comment]
@edwardhartnett - with your branch I'm still getting `NetCDF: Invalid argument` in nf90_def_var when I enable compression filters in parallel mode.  Is there something that needs to be updated in the fortran interface?

**Comment by jswhit2** [comment]
Never mind - I had LD_LIBRARY_PATH set so that it was pulling in the wrong library at runtime.

**Comment by climbfuji** [comment]
> Dom: We need a bleeding edge version of netcdf-c from https://github.com/NOAA-GSD/netcdf-c/tree/ejh_parallel_zlib compiled with parallel hdf support.

Got it, thanks.

**Comment by jswhit2** [comment]
When I run with the compression turned on, I'm getting a segfault in the ESMF library. I wonder if we need to recompile ESMF with the updated netcdf?  I think it may be pulling in the headers from the older version of netcdf that ESMF was compiled with.

**Comment by climbfuji** [comment]
I'd say that is likely.

**Comment by jswhit2** [comment]
The segfault only occurs when compression is turned on.

**Comment by climbfuji** [comment]
If you point me to your netCDF installation, I can quickly recompile ESMF for you.

**Comment by jswhit2** [comment]
/scratch2/BMC/gsienkf/whitaker/parallel-netcdf - thanks @climbfuji

**Comment by climbfuji** [comment]
It's compiling, will let you know when it is done.

**Comment by edwardhartnett** [comment]
Let me know if I can be of any assistance...

**Comment by climbfuji** [comment]
```
module use -a /scratch1/BMC/gmtb/Dom.Heinzeller/esmf-8.0.0-parallel-netcdf/modulefiles

module load intel/18.0.5.274
module load impi/2018.0.4
module load esmf/8.0.0

which nc-config
which ESMF_ESMF_Info
export | grep NETCDF
export | grep ESMFMKFILE
export | grep PATH
```

This is what the module does:

```
prereq intel/18.0.5.274
prereq impi/2018.0.4
conflict netcdf
conflict esmf

setenv              NETCDF          "/scratch2/BMC/gsienkf/whitaker/parallel-netcdf"
prepend-path        PATH            "/scratch2/BMC/gsienkf/whitaker/parallel-netcdf/bin"
prepend-path        LD_LIBRARY_PATH "/scratch2/BMC/gsienkf/whitaker/parallel-netcdf/lib"
setenv              ESMFMKFILE      "/scratch1/BMC/gmtb/Dom.Heinzeller/esmf-8.0.0-parallel-netcdf/lib/esmf.mk"
prepend-path        PATH            "/scratch1/BMC/gmtb/Dom.Heinzeller/esmf-8.0.0-parallel-netcdf/bin"
```

**Comment by jswhit2** [comment]
Thanks @climbfuji.  Unfortunately, I'm still getting the same segfault the first time parallel IO is invoked with compression on (works fine with compression off).  I think we need your help here @edwardhartnett.  Here's the traceback, for what its worth:

```
 in fcst run phase 2, na=          96
forrtl: severe (174): SIGSEGV, segmentation fault occurred
Image              PC                Routine            Line        Source
fv3.exe            0000000001D4822D  Unknown               Unknown  Unknown
libpthread-2.17.s  00002AD458ED85D0  Unknown               Unknown  Unknown
fv3.exe            0000000000AC1298  Unknown               Unknown  Unknown
fv3.exe            0000000000ABD54A  Unknown               Unknown  Unknown
fv3.exe            0000000000A2CC50  Unknown               Unknown  Unknown
libesmf.so         00002AD454BE5181  _ZN5ESMCI6FTable1        2010  ESMCI_FTable.C
libesmf.so         00002AD454BE8D66  ESMCI_FTableCallE         746  ESMCI_FTable.C
libesmf.so         00002AD4550D6EFA  _ZN5ESMCI2VM5ente        1178  ESMCI_VM.C
libesmf.so         00002AD454BE67B7  c_esmc_ftablecall         898  ESMCI_FTable.C
libesmf.so         00002AD4552FEDC1  esmf_compmod_mp_e        1209  ESMF_Comp.F90
libesmf.so         00002AD4554E9624  esmf_gridcompmod_        1889  ESMF_GridComp.F90
fv3.exe            00000000004B083A  Unknown               Unknown  Unknown
libesmf.so         00002AD4554E9624  esmf_gridcompmod_        1889  ESMF_GridComp.F90
fv3.exe            00000000004B083A  Unknown               Unknown  Unknown
libesmf.so         00002AD454F791F0  _ZN5ESMCI13Method         287  ESMCI_MethodTable.C
libesmf.so         00002AD454F79113  _ZN5ESMCI11Method         455  ESMCI_MethodTable.C
libesmf.so         00002AD454F785E6  c_esmc_methodtabl         203  ESMCI_MethodTable.C
libesmf.so         00002AD4551B4BE4  esmf_attachmethod         785  ESMF_AttachMethods.F90
libesmf.so         00002AD4559B4589  Unknown               Unknown  Unknown
libesmf.so         00002AD454BE5181  _ZN5ESMCI6FTable1        2010  ESMCI_FTable.C
libesmf.so         00002AD454BE8D66  ESMCI_FTableCallE         746  ESMCI_FTable.C
libesmf.so         00002AD4550D6EFA  _ZN5ESMCI2VM5ente        1178  ESMCI_VM.C
libesmf.so         00002AD454BE67B7  c_esmc_ftablecall         898  ESMCI_FTable.C
libesmf.so         00002AD4552FEDC1  esmf_compmod_mp_e        1209  ESMF_Comp.F90
libesmf.so         00002AD4554E9624  esmf_gridcompmod_        1889  ESMF_GridComp.F90
libesmf.so         00002AD45595EE7B  nuopc_driver_mp_r        2898  NUOPC_Driver.F90
libesmf.so         00002AD454BE5181  _ZN5ESMCI6FTable1        2010  ESMCI_FTable.C
libesmf.so         00002AD454BE8D66  ESMCI_FTableCallE         746  ESMCI_FTable.C
libesmf.so         00002AD4550D6EFA  _ZN5ESMCI2VM5ente        1178  ESMCI_VM.C
libesmf.so         00002AD454BE67B7  c_esmc_ftablecall         898  ESMCI_FTable.C
libesmf.so         00002AD4552FEDC1  esmf_compmod_mp_e        1209  ESMF_Comp.F90
libesmf.so         00002AD4554E9624  esmf_gridcompmod_        1889  ESMF_GridComp.F90
fv3.exe            00000000004A725A  Unknown               Unknown  Unknown
libesmf.so         00002AD454BE5181  _ZN5ESMCI6FTable1        2010  ESMCI_FTable.C
libesmf.so         00002AD454BE8D66  ESMCI_FTableCallE         746  ESMCI_FTable.C
libesmf.so         00002AD4550D6EFA  _ZN5ESMCI2VM5ente        1178  ESMCI_VM.C
libesmf.so         00002AD454BE67B7  c_esmc_ftablecall         898  ESMCI_FTable.C
libesmf.so         00002AD4552FEDC1  esmf_compmod_mp_e        1209  ESMF_Comp.F90
libesmf.so         00002AD454BE67B7  c_esmc_ftablecall         898  ESMCI_FTable.C
libesmf.so         00002AD4552FEDC1  esmf_compmod_mp_e        1209  ESMF_Comp.F90
libesmf.so         00002AD4554E9624  esmf_gridcompmod_        1889  ESMF_GridComp.F90
fv3.exe            000000000041BCF9  Unknown               Unknown  Unknown
fv3.exe            000000000041AABE  Unknown               Unknown  Unknown
libc-2.17.so       00002AD459D09495  __libc_start_main     Unknown  Unknown
fv3.exe            000000000041A9C5  Unknown               Unknown  Unknown
```

**Comment by climbfuji** [comment]
For what it's worth, I ran gmake check and got the following:
```
SYSTEM TESTS SUMMARY
Found 46 multi-processor system tests, 46 passed and 0 failed.

UNIT TESTS SUMMARY
Found 3476 non-exhaustive multi-processor unit tests, 3476 passed and 0 failed.
```

**Comment by climbfuji** [comment]
Line 2010 in ESMCI_FTable.C is one of those "easy to understand" pieces of code in ESMF:
```
    case FT_VOIDP4INTP: {
#if 0
std::cout << "calling out of case FT_VOIDP4INTP" << "\n";
#endif
      VoidP4IntPFunc vf = (VoidP4IntPFunc)func->funcptr;
      (*vf)((void *)comp, func->funcarg[1], func->funcarg[2],
        func->funcarg[3], userrc);
      break;
    }
```
It's the line starting with `(*vf)`. May need to ask Gerhard for help?

If it helps you (it usually doesn't help me when it comes to ESMF), I can compile ESMF again with DEBUG flags on.

**Comment by jswhit2** [comment]
Recompiled the model in debug mode and got a better traceback

```
forrtl: severe (174): SIGSEGV, segmentation fault occurred
Image              PC                Routine            Line        Source
fv3.exe            0000000001D472AD  Unknown               Unknown  Unknown
libpthread-2.17.s  00002B193BE855D0  Unknown               Unknown  Unknown
fv3.exe            0000000000AC197C  module_write_netc         581  module_write_netcdf_parallel.F90
fv3.exe            0000000000ABDC2A  module_write_netc         338  module_write_netcdf_parallel.F90
fv3.exe            0000000000A2D320  module_wrt_grid_c        1441  module_wrt_grid_comp.F90
```

**Comment by junwang-noaa** [comment]
I would guess this might not be an ESMF problem. I usually add print lines
to check where the code got seg fault. We have seen this error for many
cases, usually it's not in ESMF. I reported this kind of error report
problem to ESMF group.

**Comment by junwang-noaa** [comment]
Could it be that nbits  is zero or not defined?

**Comment by jswhit2** [comment]
Yes - I think it is something like that @junwang-noaa.  Probably nothing to do with netcdf, so don't bother looking into this yet @edwardhartnett

**Comment by jswhit2** [comment]
Works when I set nbits = 0 (lossless compression) - so it's a bug in the quantization, not the parallel write.  Should have a fix for that tomorrow.  Parallel, compressed writes are working!

**Comment by junwang-noaa** [comment]
Great! Thanks for making this working! I am going to test it on wcoss,

Jun

**Comment by climbfuji** [comment]
Great catch, Jun. Looking forward to having this capability in the model.

**Comment by edwardhartnett** [comment]
Great! Can you guys get an idea of how much faster this will be?

Also, did you try with szip? That is now also available, also with parallel writes.

**Comment by jswhit2** [comment]
At C384, with four output times/four write groups, each with 22 write tasks on hera:

Lossy compression (nbits=14)
* without parallel IO:  ~52-56 secs to write each file
* with parallel IO: ~17-20 secs

so almost 3x speedup.  Still have to check the output to see if it's correct though....

**Comment by jswhit2** [comment]
this is with zlib, have not tried szip yet.

**Comment by jswhit2** [comment]
WIth parallel IO, the file size goes up a bit

```
-rw-r--r-- 1 Jeffrey.S.Whitaker gsienkf 1282816497 Jan 10 14:12 dynf006_par_lossy.nc
-rw-r--r-- 1 Jeffrey.S.Whitaker gsienkf  947318493 Jan 10 14:01 dynf006_nopar_lossy.nc
```

uncompressed size
```
-rw-r--r-- 1 Jeffrey.S.Whitaker gsienkf 4256195968 Dec  5 15:07 dynf006_nocompress.nc
```

lossless compression
```
-rw-r--r-- 1 Jeffrey.S.Whitaker gsienkf 2948291512 Jan 10 04:28 dynf006_nopar_lossless.nc
-rw-r--r-- 1 Jeffrey.S.Whitaker gsienkf 2364584280 Jan 10 04:07 dynf006_par_lossless.nc
```

**Comment by edwardhartnett** [comment]
Awesome! The size going up a bit is expected behavior when using zlib with parallel I/O.

Not sure what branch you built, but my most recent PR to the netcdf-c repo (not yet merged) provides new C function nc_def_var_szip(). You probably need that in Fortran, right? I will submit a PR to the netcdf-fortran repo with that so you can easily test szip with parallel I/O.

(Also you must build HDF5 so that it finds libsz when it is built.)

**Comment by jswhit2** [comment]
Ed - that would be great.  Yes we do need it in Fortran.

**Comment by jswhit2** [comment]
The difference in compression isn't as large as I thought. I was using shuffle filter in the parallel case, and no shuffle in the serial case.  With shuffle off in both I get

```
-rw-r--r-- 1 Jeffrey.S.Whitaker gsienkf 947318493 Jan 10 14:01 dynf006_nopar_lossy.nc
-rw-r--r-- 1 Jeffrey.S.Whitaker gsienkf 951805414 Jan 10 15:48 dynf006_par_lossy.nc
```

so only a very slight difference.

**Comment by junwang-noaa** [comment]
I have trouble to compile the netcdf-c on mars. I compiled the library
under emc.nemspara, only three modules are loaded:

 1) EnvVars/1.0.3   2) ips/18.0.1.163   3) impi/18.0.1

in the config.log, I have:
/gpfs/dell2/emc/modeling/noscrub/emc.nemspara/soft/netcdf_parallel/include/H5public.h:63:20:
fatal error: mpi.h: No such file or directory
 #   include <mpi.h>
                    ^
compilation terminated.

Since I do have impi loaded, is there something else required to
compile netcdf-c?

 hdf5-1.10.6 was installed successfully.

**Comment by edwardhartnett** [comment]
Did you try setting CC=mpicc

**Comment by CoryMartin-NOAA** [comment]
When I compiled the release version on Mars I used
CC=mpiicc
FC=mpiifort

**Comment by jswhit2** [comment]
Just added a pull request (PR #41) for this so it's easier for people to examine the code changes.

**Comment by jswhit2** [comment]
Output for lossless compression is identical with and without parallel IO.

For lossy compression, there are slight differences.  Not sure if this is to be expected or not.

**Comment by CoryMartin-NOAA** [comment]
@jswhit2 could the differences for lossy compression be due to the way the quantizing is done and not the I/O?

**Comment by jswhit2** [comment]
Yes - I think it may be an order of operations thing.  The difference is about the same as the quantization error itself.

**Comment by jswhit2** [comment]
BTW, here's what I use to check the difference in output files

```python
from __future__ import print_function
from netCDF4 import Dataset
import numpy as np
import sys
nc1 = Dataset(sys.argv[1])
nc2 = Dataset(sys.argv[2])
for varname in nc1.variables.keys():
    data1 = nc1[varname][:]
    data2 = nc2[varname][:]
    diff = data2-data1
    print('%s min/max 1=%s,%s min/max 2=%s,%s max abs diff=%s'%(varname, data1.min(), data2.max(), data2.min(), data2.max(), (np.abs(diff)).max()))
```

**Comment by jswhit2** [comment]
After re-ordering the quantization calculation in the parallel version to match the serial version, the output is now identical.

**Comment by jswhit2** [comment]
Scaling tests.  C384, 4 output times, 432 compute tasks, 4 write groups  
Seems like at this resolution there's no point in having more than 12 write tasks per group.

```
* 2 write tasks per write group (total 440 tasks)
 parallel netcdf      Write Time is   31.03578 at Fcst   00:00
 parallel netcdf      Write Time is   11.42732 at Fcst   00:00
 total            Write Time is   43.97696 at Fcst   00:00
 parallel netcdf      Write Time is   30.44723 at Fcst   03:00
 parallel netcdf      Write Time is   11.22503 at Fcst   03:00
 total            Write Time is   43.18365 at Fcst   03:00
 parallel netcdf      Write Time is   31.32110 at Fcst   06:00
 parallel netcdf      Write Time is   11.43270 at Fcst   06:00
 total            Write Time is   44.25961 at Fcst   06:00
* 4 write tasks per write group (total 448 tasks)
 total            Write Time is   26.89187 at Fcst   00:00
 parallel netcdf      Write Time is   15.31839 at Fcst   03:00
 parallel netcdf      Write Time is   10.79294 at Fcst   03:00
 total            Write Time is   26.86619 at Fcst   03:00
 parallel netcdf      Write Time is   15.29695 at Fcst   06:00
 parallel netcdf      Write Time is   10.81599 at Fcst   06:00
 total            Write Time is   26.86715 at Fcst   06:00
 parallel netcdf      Write Time is   14.79226 at Fcst   09:00
 parallel netcdf      Write Time is   10.47566 at Fcst   09:00
* 8 write tasks per write group (total 464 tasks)
 total            Write Time is   19.91427 at Fcst   00:00
 parallel netcdf      Write Time is    8.47844 at Fcst   03:00
 parallel netcdf      Write Time is   10.80266 at Fcst   03:00
 total            Write Time is   19.66060 at Fcst   03:00
 parallel netcdf      Write Time is    8.68719 at Fcst   06:00
 parallel netcdf      Write Time is   10.78251 at Fcst   06:00
 total            Write Time is   19.83424 at Fcst   06:00
 parallel netcdf      Write Time is    7.48860 at Fcst   00:00
 parallel netcdf      Write Time is   10.81924 at Fcst   00:00
* 12 write tasks per write group (total 480 tasks)
 total            Write Time is   18.69116 at Fcst   00:00
 parallel netcdf      Write Time is    7.10439 at Fcst   03:00
 parallel netcdf      Write Time is   10.67810 at Fcst   03:00
 total            Write Time is   18.02822 at Fcst   03:00
 parallel netcdf      Write Time is    6.93121 at Fcst   06:00
 parallel netcdf      Write Time is   10.66627 at Fcst   06:00
 total            Write Time is   17.84311 at Fcst   06:00
 parallel netcdf      Write Time is    8.50981 at Fcst   09:00
 parallel netcdf      Write Time is   10.67957 at Fcst   09:00
* 22 write tasks per write group (total 520 tasks)
 parallel netcdf      Write Time is    6.36622 at Fcst   00:00
 parallel netcdf      Write Time is   11.07094 at Fcst   00:00
 total            Write Time is   17.61161 at Fcst   00:00
 parallel netcdf      Write Time is    9.21893 at Fcst   03:00
 parallel netcdf      Write Time is   11.64303 at Fcst   03:00
 total            Write Time is   20.99851 at Fcst   03:00
 parallel netcdf      Write Time is    6.22021 at Fcst   06:00
 parallel netcdf      Write Time is   10.92823 at Fcst   06:00
 total            Write Time is   17.28535 at Fcst   06:00
 parallel netcdf      Write Time is    9.49604 at Fcst   09:00
 parallel netcdf      Write Time is   10.95753 at Fcst   09:00
```

For reference, the result without parallel IO
```
* 2 write tasks per write group (total 440 tasks)
 netcdf      Write Time is   56.08578 at Fcst   00:00
 netcdf      Write Time is    9.19766 at Fcst   00:00
 total            Write Time is   66.77824 at Fcst   00:00
 netcdf      Write Time is   56.22830 at Fcst   03:00
 netcdf      Write Time is    8.34209 at Fcst   03:00
 total            Write Time is   66.06653 at Fcst   03:00
 netcdf      Write Time is   57.31763 at Fcst   06:00
 netcdf      Write Time is    8.61243 at Fcst   06:00
 total            Write Time is   67.42009 at Fcst   06:00
* 12 write tasks per write group (total 480 tasks)
 total            Write Time is   19.42652 at Fcst   09:00
 netcdf      Write Time is   53.54677 at Fcst   00:00
 netcdf      Write Time is    8.56202 at Fcst   00:00
 total            Write Time is   62.46376 at Fcst   00:00
 netcdf      Write Time is   52.60850 at Fcst   03:00
 netcdf      Write Time is    8.22610 at Fcst   03:00
 total            Write Time is   61.08010 at Fcst   03:00
 netcdf      Write Time is   53.07427 at Fcst   06:00
 netcdf      Write Time is    8.44276 at Fcst   06:00
 total            Write Time is   61.77053 at Fcst   06:00
* 22 write tasks per write group (total 520 tasks)
 netcdf      Write Time is   52.59507 at Fcst   00:00
 netcdf      Write Time is    8.18125 at Fcst   00:00
 total            Write Time is   60.95557 at Fcst   00:00
 netcdf      Write Time is   52.17142 at Fcst   03:00
 netcdf      Write Time is    8.15561 at Fcst   03:00
 total            Write Time is   60.46364 at Fcst   03:00
 netcdf      Write Time is   56.41519 at Fcst   06:00
 netcdf      Write Time is    8.21870 at Fcst   06:00
 total            Write Time is   64.77179 at Fcst   06:00
```

**Comment by jswhit2** [comment]
All tests so far have been done with the global model - need to test FV3SAR also.

**Comment by junwang-noaa** [comment]
I am using CC=mpiicc, and FC=mpiifort, the netcdf-c got compiled. When I compile netcdf-fortran-4.5.2, I got error:

configure:23833: mpiicc -o conftest -g -O2 -I/gpfs/dell2/emc/modeling/noscrub/emc.nemspara/soft/netcdf_parallel/include -L/gpfs/dell2/emc/modeling/noscrub/emc.nemspara/soft/netcdf_parallel/lib conftest.c -lnetcdf -lm  >&5
configure:23833: $? = 0
configure:23833: ./conftest
./conftest: error while loading shared libraries: libnetcdf.so.15: cannot open shared object file: No such file or directory
configure:23833: $? = 127

however, the libnetcdf.so.15 exists under /gpfs/dell2/emc/modeling/noscrub/emc.nemspara/soft/netcdf_parallel/lib

[emc.nemspara@m72a2 lib]$ ls -l /gpfs/dell2/emc/modeling/noscrub/emc.nemspara/soft/netcdf_parallel/lib
total 25088
...
lrwxrwxrwx 1 emc.nemspara emcmodel       19 Jan 10 17:51 libnetcdf.so -> libnetcdf.so.15.2.1
lrwxrwxrwx 1 emc.nemspara emcmodel       19 Jan 10 17:51 libnetcdf.so.15 -> libnetcdf.so.15.2.1
-rwxr-xr-x 1 emc.nemspara emcmodel  1418176 Jan 10 17:51 libnetcdf.so.15.2.1
...

Any idea what might be wrong?

**Comment by CoryMartin-NOAA** [comment]
@junwang-noaa looks like /gpfs/dell2/emc/modeling/noscrub/emc.nemspara/soft/netcdf_parallel/lib needs to be added to your LD_LIBRARY_PATH. It's there for compiling but not available at runtime for ./conftest.

**Comment by edwardhartnett** [comment]
@jswhit adding szip to netcdf-fortran is going to require some negotiations. However, if you build the current netcdf-fortran against the netcdf-c lib you built, you can use nf_def_var_filter() to turn on the szip filter, and it will work in parallel I/O.

**Comment by junwang-noaa** [comment]
Thanks, Cori. I am able to get the netcdf lib installed on mars.

@jswhit2 do you create netcdf/hfd5 module files or you just define the environment variables in the  modulefiles/hera.intel/fv3? Do you have a ufs-weather-model branch ready for the parallel writing test?

**Comment by jswhit2** [comment]
@junwang-noaa: I have not yet created module files (I just set the env vars - see `/scratch2/BMC/gsienkf/whitaker/ufs-weather-model-parnc/modulefiles/hera.intel/fv3`).  At runtime all you need is 

```
module purge
module load intel/18.0.5.274
module load impi/2018.0.4
setenv LD_LIBRARY_PATH "/scratch2/BMC/gsienkf/whitaker/parallel-netcdf/lib:${LD_LIBRARY_PATH}"
```

It should be ready for testing, although it still needs modulefiles created on hera and WCOSS, plus a mechanism for turning off parallel netcdf support at build time.

**Comment by junwang-noaa** [comment]
I am able to run some test on wcoss dell. Here are some results from C192 with 1 write group and 12 tasks per group:

Netcdf compressed: ideflate=1, nbits=14, 
sequential:
[288]  netcdf      Write Time is   13.24216 at Fcst   24:00
[288]  netcdf      Write Time is    3.08769 at Fcst   24:00
[288]  total            Write Time is   16.41797 at Fcst   24:00
parallel:
[288]  parallel netcdf      Write Time is    3.96522 at Fcst   00:15
[288]  parallel netcdf      Write Time is    4.49794 at Fcst   00:15
[288]  total            Write Time is    8.55112 at Fcst   00:15

From the test, it looks to me parallel writing the 3D model fields is about 1/3 of the sequential writing.  But the time of writing the 2D surface fields increased. The results are consistent with the results from Jeff's run on c384.  But I am curious why parallel writing of 2D does not speed up.

**Comment by jswhit2** [comment]
I think that there just isn't enough data being written to overcome the communications overhead for the 2d surface fields.

**Comment by junwang-noaa** [comment]
So for parallel writing compressed data, is it possible that we can do parallel writing on the number of fields, e.g. writing slides of data (istt:iend, jstt:jend,nfield_stt:nfield_end)?

**Comment by junwang-noaa** [comment]
Here are some test results on wcoss dell with C384 fv3 ensemble run. The data size is slightly increased.

parallel writing:
-rw-r--r-- 1 Jun.Wang g01 1844435007 Jan 22 03:27 atmf000.nc
-rw-r--r-- 1 Jun.Wang g01 1830870160 Jan 22 03:29 atmf001.nc
-rw-r--r-- 1 Jun.Wang g01  279801964 Jan 22 03:28 sfcf000.nc
-rw-r--r-- 1 Jun.Wang g01  270925570 Jan 22 03:30 sfcf001.nc

No parallel:
-rw-r--r-- 1 Jun.Wang g01 1840588536 Jan 21 16:39 atmf000.nc
-rw-r--r-- 1 Jun.Wang g01 1819256319 Jan 21 16:41 atmf001.nc
-rw-r--r-- 1 Jun.Wang g01  276932033 Jan 21 16:40 sfcf000.nc
-rw-r--r-- 1 Jun.Wang g01  268183050 Jan 21 16:41 sfcf001.nc

using current setting 7 write group, 28 tasks per group
Parallel writing:
[384]  parallel netcdf      Write Time is   24.28754 at Fcst   00:00
[384]  parallel netcdf      Write Time is   84.23861 at Fcst   00:00
[384]  total            Write Time is  131.68151 at Fcst   00:00
[412]  parallel netcdf      Write Time is   21.73747 at Fcst   01:00
[412]  parallel netcdf      Write Time is   81.98856 at Fcst   01:00
[412]  total            Write Time is  129.67429 at Fcst   01:00

No parallel writing:
 netcdf      Write Time is  118.42970 at Fcst   00:00
 netcdf      Write Time is   12.19001 at Fcst   00:00
 total            Write Time is  153.53147 at Fcst   00:00
 netcdf      Write Time is  109.14027 at Fcst   01:00
 netcdf      Write Time is   10.60381 at Fcst   01:00
 total            Write Time is  145.54639 at Fcst   01:00
 netcdf      Write Time is  109.61039 at Fcst   02:00
 netcdf      Write Time is   10.92568 at Fcst   02:00
 total            Write Time is  146.59644 at Fcst   02:00

Parallel writing 3D fields reduces time from 110s to 24s, but parallel writing 2D fields increases from 10s to 80s. Does anybody have any idea why?

**Comment by junwang-noaa** [comment]
Here are results from C768L127, results look promising. Writing the 3D file atmf file speeds up ~15 times (450s down to 30s), while 2D surface file writing increases 60% (45s to 70s). Total time on writing grid comp (writing 2 history files and inline post) for one forecast hour is reduced from 550s->150s. Inline post takes 50s in both runs.  

file size:
parallel writing:
-rw-r--r-- 1 Jun.Wang g01 6680866372 Jan 22 11:45 atmf000.nc
-rw-r--r-- 1 Jun.Wang g01 6769628821 Jan 22 11:48 atmf001.nc
-rw-r--r-- 1 Jun.Wang g01 6675425724 Jan 22 11:51 atmf002.nc
-rw-r--r-- 1 Jun.Wang g01 1060341025 Jan 22 11:46 sfcf000.nc
-rw-r--r-- 1 Jun.Wang g01 1026448059 Jan 22 11:49 sfcf001.nc
-rw-r--r-- 1 Jun.Wang g01 1044235908 Jan 22 11:52 sfcf002.nc

No parallel writing:
-rw-r--r-- 1 Jun.Wang g01 6677616851 Jan 21 22:29 atmf000.nc
-rw-r--r-- 1 Jun.Wang g01 6778890732 Jan 21 22:31 atmf001.nc
-rw-r--r-- 1 Jun.Wang g01 6663228925 Jan 21 22:33 atmf002.nc
-rw-r--r-- 1 Jun.Wang g01 1049063514 Jan 21 22:30 sfcf000.nc
-rw-r--r-- 1 Jun.Wang g01 1015687585 Jan 21 22:32 sfcf001.nc
-rw-r--r-- 1 Jun.Wang g01 1033478036 Jan 21 22:34 sfcf002.nc

Parallel:
 parallel netcdf      Write Time is   33.48398 at Fcst   00:00
 parallel netcdf      Write Time is   69.58479 at Fcst   00:00
 parallel netcdf      Write Time is   30.34695 at Fcst   01:00
 parallel netcdf      Write Time is   70.54954 at Fcst   01:00
 parallel netcdf      Write Time is   29.85357 at Fcst   02:00
 parallel netcdf      Write Time is   71.57624 at Fcst   02:00

No parallel writing:
 netcdf      Write Time is  426.01005 at Fcst   00:00
 netcdf      Write Time is   43.87808 at Fcst   00:00
 netcdf      Write Time is  454.70615 at Fcst   01:00
 netcdf      Write Time is   48.44027 at Fcst   01:00
 netcdf      Write Time is  448.26954 at Fcst   02:00
 netcdf      Write Time is   46.95907 at Fcst   02:00

**Comment by edwardhartnett** [comment]
OK, szip write capability is waiting a merge to master on the netcdf-c project. Once that is done, you can try these tests using szip as well as zlib. (Or you can build netcdf-c branch ejh_szip).

I am also working on making the LZ4 compression filter available. Reportedly it is a lot faster. But that's not ready for you to try yet...

**Comment by jswhit** [comment]
@junwang-noaa - I'm looking to turn off parallel writes for the 2d file in model_wrt_grid_comp.F90.  I can do this by checking the value of nbdl (nbdl == 1 is the 3D file).  Is there a better way to check which bundle contains the 3D fields?

**Comment by jswhit** [comment]
I went ahead and turned off parallel writes for the 2d file - this should provide another 25 second gain in the C768 test.

**Comment by jswhit** [comment]
I'm wary of szip and lz4, unless we really need to wring out some more speed for compressed writes.  It will make it much harder to end users - they will have to rebuild  the  hdf5 and netcdf  libs to enable the extra filters.  zlib is available in the default builds.

**Comment by junwang-noaa** [comment]
Jeff,

The surface file can actually have 3D fields such as radar reflectivity, or
physics 3D tendency fields, just in current setting we do not output the 3D
fields in sfc. I would suggest to use a logic variable to mark if an output
bundle has 3D fields, if it is not we can using sequential write. We can
set the logic variable at the initialization step.

I can test the code with non-parallel writing for sfc file with all 2D
fields tomorrow when I get access to wcoss.

**Comment by jswhit2** [comment]
Or we could make `output_file` an array (like `filename_base`) so we control whether parallel IO is used on a per-file basis, e.g:

```
num_files:               2
filename_base:           'dyn' 'phy'
output_grid:             'gaussian_grid'
output_file:             'netcdf_parallel' 'netcdf'
```

**Comment by edwardhartnett** [comment]
WRT what type of compression to use, there are now several options.

Zlib is built in to every netCDF-4 install, so is always present. However, it is slow for writing and reading.

Szlib has been supported for read access since the netcdf-c-4.0 release, but, in the next release, will also be supported for writing. Users must have built HDF5 with szlib or they will get an error when they try to read the data (NC_EFILTER). Readers that don't currently support szlib will need to rebuild HDF5 and netcdf-c to get support.

The other new compression methods do not require a rebuild of HDF5, they just require the installation of the Community Codec Repo for netCDF. The CCR release includes the filter code and will install the filters in the correct place for HDF5 to find them. Recompile of HDF5 and netCDF is not required. Readers of these data will also have to install the CCR project, then their read code will work transparently.

LZ4 is a particularly promising compression method, which promises to be much faster on writes and reads. Unidata has indicated that if NOAA were to start using LZ4 compression, then Unidata will consider packaging the LZ4 filter with all releases, and the CCR would not be required in that case.

To summarize, netcdf support for LZ4 compression is good and will improve. The CCR project will allow us to ensure that all readers of the data have easy access to uncompression filters.

**Comment by junwang-noaa** [comment]
Jeff,

That will also work and give us some flexibility for low resolution runs.

Jun

**Comment by jswhit2** [comment]
Ed - thanks for the clarification. Sounds like it's easier than I thought to use non-zlib compression filters.

Jun - I went ahead implemented `output_file` as an array so we can specify it differently for 3d and 2d files.  It works for `netcdf` and `netcdf_parallel`, but it won't work if you try to mix `nemsio` and `netcdf_parallel`  since they both will try to use the same MPI communicator.  Any suggestions on how to deal with this?

**Comment by junwang-noaa** [comment]
Can we check if  output_file(1) has "netcdf" substring, all the elements in
output_file need to have "netcdf", vice versa.

**Comment by jswhit** [comment]
pushed a change to  https://github.com/junwang-noaa/fv3atm/tree/netcdf_parallel to allow 

```
num_files:               2
filename_base:           'dyn' 'phy'
output_grid:             'gaussian_grid'
output_file:             'netcdf_parallel' 'netcdf'
```

A single value for output_file will still work.  If multiple values are given, they all must start with 'netcdf' (or else the model will abort with an error).

**Comment by jswhit2** [comment]
Using the default chunksize for 2D variables eliminates the slowdown - with this change using parallel IO for the 2D file is faster than serial IO.

**Comment by junwang-noaa** [comment]
That's great news, thanks, Jeff!

**Comment by jswhit** [comment]
With these changes, here are the benchmarks on hera
C384, 4 output times, 432 compute tasks, 4 write groups, 12 write tasks per group (total 480 cores)

serial IO (with compression)
```
 netcdf      Write Time is   55.37413 at Fcst   00:00
 netcdf      Write Time is    8.66331 at Fcst   00:00
 total            Write Time is   64.28766 at Fcst   00:00
 netcdf      Write Time is   54.16932 at Fcst   03:00
 netcdf      Write Time is    8.44123 at Fcst   03:00
 total            Write Time is   62.85684 at Fcst   03:00
 netcdf      Write Time is   55.36696 at Fcst   06:00
 netcdf      Write Time is    8.54700 at Fcst   06:00
 total            Write Time is   64.16426 at Fcst   06:00
 netcdf      Write Time is   53.03024 at Fcst   09:00
 netcdf      Write Time is    8.42973 at Fcst   09:00
 total            Write Time is   61.74685 at Fcst   09:00
The total amount of wall time                        = 518.930422
```

parallel IO (with compression)
```
 parallel netcdf      Write Time is    6.99810 at Fcst   00:00
 parallel netcdf      Write Time is    3.61901 at Fcst   00:00
 total            Write Time is   10.91470 at Fcst   00:00
 parallel netcdf      Write Time is    6.73620 at Fcst   03:00
 parallel netcdf      Write Time is    3.50100 at Fcst   03:00
 total            Write Time is   10.48364 at Fcst   03:00
 parallel netcdf      Write Time is    6.98989 at Fcst   06:00
 parallel netcdf      Write Time is    3.51980 at Fcst   06:00
 total            Write Time is   10.75907 at Fcst   06:00
 parallel netcdf      Write Time is    8.14131 at Fcst   09:00
 parallel netcdf      Write Time is    3.79231 at Fcst   09:00
 total            Write Time is   12.25741 at Fcst   09:00
The total amount of wall time                        = 456.915771
```

serial IO (no compression)
```
 netcdf      Write Time is    9.57775 at Fcst   00:00
 netcdf      Write Time is    1.13526 at Fcst   00:00
 total            Write Time is   11.00683 at Fcst   00:00
 netcdf      Write Time is    7.43096 at Fcst   03:00
 netcdf      Write Time is    1.14806 at Fcst   03:00
 total            Write Time is    8.82465 at Fcst   03:00
 netcdf      Write Time is    7.62321 at Fcst   06:00
 netcdf      Write Time is    1.14155 at Fcst   06:00
 total            Write Time is    9.01934 at Fcst   06:00
 netcdf      Write Time is    8.14069 at Fcst   09:00
 netcdf      Write Time is    1.30981 at Fcst   09:00
 total            Write Time is    9.91153 at Fcst   09:00
The total amount of wall time                        = 459.836954
```

parallel IO (no compression)
```
 parallel netcdf      Write Time is    5.44953 at Fcst   00:00
 parallel netcdf      Write Time is    0.79031 at Fcst   00:00
 total            Write Time is    6.54253 at Fcst   00:00
 parallel netcdf      Write Time is    3.59653 at Fcst   03:00
 parallel netcdf      Write Time is    1.11380 at Fcst   03:00
 total            Write Time is    4.95609 at Fcst   03:00
 parallel netcdf      Write Time is    3.10670 at Fcst   06:00
 parallel netcdf      Write Time is    0.91198 at Fcst   06:00
 total            Write Time is    4.27648 at Fcst   06:00
 parallel netcdf      Write Time is    5.28217 at Fcst   09:00
 parallel netcdf      Write Time is    1.17067 at Fcst   09:00
 total            Write Time is    6.76937 at Fcst   09:00
The total amount of wall time                        = 464.404551
```

**Comment by edwardhartnett** [comment]
All the parallel I/O write times are faster. Why is the wall time slower?

**Comment by jswhit** [comment]
Sorry, I had those total wall times backwards.  Fixed now.

**Comment by edwardhartnett** [comment]
I still don't understand why all the wall times are so similar. When I look at the numbers for the write times, I see parallel I/O is a lot faster, but the wall clock times are very similar.

For example:, serial with compression:

 total            Write Time is   64.28766 at Fcst   00:00

Parallel with compression:

 total            Write Time is   10.91470 at Fcst   00:00

So it has gone down by a factor of 6. And all the other times similar. So why is the wall time not down by a factor of 6 as well?

**Comment by jswhit2** [comment]
Ed- the wall time is for the entire model forecast, the other timings are just for the IO.  Plus, the IO is asynchronous and runs on separate MPI tasks, so the model doesn't need to wait for the IO to finish.

**Comment by jswhit2** [comment]
I added an option to build without parallel netcdf support using `-DNO_PARALLEL_NETCDF `using a stub (`module_write_netcdf_parallel_stub.F90`).  

Should be ready for merge now.

**Comment by junwang-noaa** [comment]
Some results from wcoss dell with C768L127  compressed history files. Using parallel writing sfc file with default chunk size increases the  

Using arallel writing for 3D fields and parallel writing with default chunksize for surface file:
parallel netcdf      Write Time is   32.28131 at Fcst   00:00
 parallel netcdf      Write Time is   98.26427 at Fcst   00:00
 total            Write Time is  184.96057 at Fcst   00:00
 parallel netcdf      Write Time is   30.95869 at Fcst   01:00
 parallel netcdf      Write Time is   97.88160 at Fcst   01:00
 total            Write Time is  181.04889 at Fcst   01:00
 parallel netcdf      Write Time is   36.54861 at Fcst   02:00
 parallel netcdf      Write Time is   98.68027 at Fcst   02:00
 total            Write Time is  187.47505 at Fcst   02:00
 parallel netcdf      Write Time is   31.66576 at Fcst   03:00
 parallel netcdf      Write Time is   97.60119 at Fcst   03:00
 total            Write Time is  182.98560 at Fcst   03:00

Parallel writing with default chunk size increased the writing time.
-                    shuffle=.true.,deflate_level=ideflate,&
-                    chunksizes=(/im,jm,1/),cache_size=40*im*jm); NC_ERR_STOP(ncerr)
+                    shuffle=.true.,deflate_level=ideflate); NC_ERR_STOP(ncerr)
previous run with chunksize /im,jm,1/:
parallel netcdf      Write Time is   33.48398 at Fcst   00:00
 parallel netcdf      Write Time is   69.58479 at Fcst   00:00
 total            Write Time is  157.82494 at Fcst   00:00
 parallel netcdf      Write Time is   30.34695 at Fcst   01:00
 parallel netcdf      Write Time is   70.54954 at Fcst   01:00
 total            Write Time is  153.16201 at Fcst   01:00
 parallel netcdf      Write Time is   29.85357 at Fcst   02:00
 parallel netcdf      Write Time is   71.57624 at Fcst   02:00
 total            Write Time is  153.87252 at Fcst   02:00
 parallel netcdf      Write Time is   31.30396 at Fcst   03:00
 parallel netcdf      Write Time is   72.84347 at Fcst   03:00
 total            Write Time is  158.10736 at Fcst   03:00

Using parallel writing for 3D fields and sequential writing for sfc file, 
parallel netcdf      Write Time is   31.68672 at Fcst   00:00
 netcdf      Write Time is   49.13957 at Fcst   00:00
 total            Write Time is  136.93367 at Fcst   00:00
 parallel netcdf      Write Time is   33.27543 at Fcst   01:00
 netcdf      Write Time is   45.69605 at Fcst   01:00
 total            Write Time is  131.90097 at Fcst   01:00
 parallel netcdf      Write Time is   32.26719 at Fcst   02:00
 netcdf      Write Time is   46.24452 at Fcst   02:00
 total            Write Time is  131.46787 at Fcst   02:00
 parallel netcdf      Write Time is   29.84425 at Fcst   03:00
 netcdf      Write Time is   46.99665 at Fcst   03:00
 total            Write Time is  131.50041 at Fcst   03:00

So I think for fv3 parallel, we can use the parallel writing for 3D history files, while sequential writing for 2D history files

**Comment by junwang-noaa** [comment]
I also listed the total time spending on the write grid component(including time writing history files and inline post), it is actually decreasing when using parallel writing. The total run time is decided by the forecast time in this case (C768L127). So far I saw several issues:

1) writing one time level restart file by forecast tasks takes 
 PASS: fcstRUN phase 2, na =          143  time is    291.485043048859
2) at every writing time, it seems there is some extra time on communication that impact the forecast time in 1 or 2 time steps:
...
PASS: fcstRUN phase 1, na =           23  time is    5.95756697654724
 PASS: fcstRUN phase 2, na =           23  time is   0.268661975860596
 PASS: fcstRUN phase 1, na =           24  time is    **8.62026309967041**
 PASS: fcstRUN phase 2, na =           24  time is    1.23025012016296
 PASS: fcstRUN phase 1, na =           25  time is    **13.1568639278412**
 PASS: fcstRUN phase 2, na =           25  time is   0.207802057266235
...
PASS: fcstRUN phase 1, na =           47  time is    5.95676708221436
 PASS: fcstRUN phase 2, na =           47  time is   0.268540859222412
 PASS: fcstRUN phase 1, na =           48  time is    **8.67054295539856**
 PASS: fcstRUN phase 2, na =           48  time is    1.20829796791077
 PASS: fcstRUN phase 1, na =           49  time is    **13.2787740230560**
 PASS: fcstRUN phase 2, na =           49  time is   0.188354969024658
 PASS: fcstRUN phase 1, na =           50  time is    5.96231389045715
 PASS: fcstRUN phase 2, na =           50  time is   0.189077854156494

To finish one hour forecast, forecast time needs 24*6.2 ~150, the total writing time is: 140s, so we actually only need 2 write groups, instead of current 6 groups. On the other side, we need to improve the forecast time.

**Comment by jswhit2** [comment]
I ran a C768 test on hera, and I still see a benefit to parallel IO for the 2D files.  Using 12 write tasks, with parallel IO for both 2d and 3d files I get

```
 parallel netcdf      Write Time is   36.91967 at Fcst   03:00
 parallel netcdf      Write Time is   18.39263 at Fcst   03:00
 total                Write Time is   55.56771 at Fcst   03:00
```

whereas turning off parallel IO for 2d files I see

```
 parallel netcdf      Write Time is   37.76248 at Fcst   03:00
 netcdf               Write Time is   33.08457 at Fcst   03:00
 total                Write Time is   71.07401 at Fcst   03:00
```

and without parallel IO for either 2d or 3d files

```
 netcdf            Write Time is  206.89221 at Fcst   03:00
 netcdf            Write Time is   29.70651 at Fcst   03:00
 total             Write Time is  236.85511 at Fcst   03:00
```

**Comment by jswhit2** [comment]
Seems like the 2D writes are very sensitive to chunksize.  If I set the chunksize to be the same as the size of the array on each write task, I get

```
 parallel netcdf      Write Time is   35.68702 at Fcst   03:03
 parallel netcdf      Write Time is    9.80075 at Fcst   03:03
 total                Write Time is   45.72166 at Fcst   03:03
```

The optimal chunksize may be platform dependent.  I'll look at adding the chunksize as a runtime parameter in model_configure

**Comment by jswhit2** [comment]
Added ichunk2d, jchunk2d parameters in model_configure (can be used to tune chunksize for best parallel IO performance).  Default is size of array on each write task.

**Comment by jswhit2** [comment]
Also added ichunk3d,jchunk3d,kchunk3d to set 3d variable chunksize.  Default is ichunk3d=ichunk2d,  jchunk3d=jchunk2d, kchunk3d=nlevs.  This results in the fastest writes for me on hera:

```
 parallel netcdf      Write Time is   24.97020 at Fcst   03:03
 parallel netcdf      Write Time is    9.98413 at Fcst   03:03
 total                Write Time is   35.23754 at Fcst   03:03
```

To restore the previous behavior, set ichunk3d=imo,jchunk3d=jmo,kchunk3d=1.

Writes are slower for this setting, but reading 2d horizontal slices is much faster.

@junwang-noaa  - could you run your test again on WCOSS with these new default chunksizes?

**Comment by junwang-noaa** [comment]
Here are results using default chunksize (decomposition size on write
tasks).

parallel netcdf      Write Time is   43.48440 at Fcst   00:00
 parallel netcdf      Write Time is   42.56229 at Fcst   00:00
 total            Write Time is  141.08820 at Fcst   00:00
 parallel netcdf      Write Time is   45.74817 at Fcst   01:00
 parallel netcdf      Write Time is   29.93798 at Fcst   01:00
 total            Write Time is  128.55868 at Fcst   01:00
 parallel netcdf      Write Time is   37.90728 at Fcst   02:00
 parallel netcdf      Write Time is   42.14662 at Fcst   02:00
 total            Write Time is  131.89296 at Fcst   02:00
 parallel netcdf      Write Time is   50.03225 at Fcst   03:00
 parallel netcdf      Write Time is   39.72588 at Fcst   03:00
 total            Write Time is  141.78540 at Fcst   03:00

Using chunksize  3D:(imo,jmo,1,nlevs), 2D:(imo,jmo,1)

parallel netcdf Write Time is 31.30396 at Fcst 03:00
parallel netcdf Write Time is 72.84347 at Fcst 03:00
total Write Time is 158.10736 at Fcst 03:00

Using 3D chunk size (imo,jmo,1,nlevs), sequential write 2D:

parallel netcdf Write Time is 29.84425 at Fcst 03:00
netcdf Write Time is 46.99665 at Fcst 03:00
total Write Time is 131.50041 at Fcst 03:00

**Comment by jswhit2** [comment]
did you mean (imo,jmo,nlevs,1) for the 3d chunk size?

**Comment by junwang-noaa** [comment]
Sorry,I am using an old version, I think it is (im,jm,1,1). It looks to me
writing 3D fields with  default chunksize is slower on dell compared to
previous version.

@@ -176,7 +173,7 @@ module module_write_netcdf_parallel
             ncerr = nf90_def_var(ncid, trim(fldName), NF90_FLOAT, &
                     (/im_dimid,jm_dimid,pfull_dimid,time_dimid/),
varids(i), &
                     shuffle=.false.,deflate_level=ideflate,&
-                    chunksizes=(/im,jm,1,1/),cache_size=40*im*jm);
NC_ERR_STOP(ncerr)
+                    chunksizes=(/ichunk3d,jchunk3d,kchunk3d,1/));
NC_ERR_STOP(ncerr)
             ! compression filters require collective access.

**Comment by jswhit2** [comment]
Hmm.  Wonder why I'm getting ~4x faster writes on hera with the new default chunksizes and 12 write tasks (total write time of 35 secs vs 140 secs on WCOSS).

**Comment by junwang-noaa** [comment]
Jeff,

The 140s on dell includes 50s inline post time. So the total time is about
90s for writing netcdf files.

**Comment by junwang-noaa** [comment]
Here are results on dell using default chunk size for 2D file, and
(im,jm,1,1) for 3D files, basically writing netcdf files takes 60-70s. In
this run, I am using 2 write groups instead of 7 groups used in the initial
test from fv3_parallel (reducing 300 tasks).

parallel netcdf      Write Time is   32.15252 at Fcst   00:00
 parallel netcdf      Write Time is   45.53429 at Fcst   00:00
 total            Write Time is  132.93790 at Fcst   00:00
 parallel netcdf      Write Time is   31.60762 at Fcst   01:00
 parallel netcdf      Write Time is   43.13410 at Fcst   01:00
 total            Write Time is  127.48863 at Fcst   01:00
 parallel netcdf      Write Time is   30.07048 at Fcst   02:00
 parallel netcdf      Write Time is   33.14977 at Fcst   02:00
 total            Write Time is  114.95780 at Fcst   02:00
 parallel netcdf      Write Time is   32.54118 at Fcst   03:00
 parallel netcdf      Write Time is   37.68903 at Fcst   03:00
 total            Write Time is  122.18182 at Fcst   03:00
 parallel netcdf      Write Time is   30.96448 at Fcst   04:00
 parallel netcdf      Write Time is   31.85908 at Fcst   04:00
 total            Write Time is  114.86909 at Fcst   04:00
 parallel netcdf      Write Time is   32.27664 at Fcst   05:00
 parallel netcdf      Write Time is   25.77598 at Fcst   05:00
 total            Write Time is  110.33424 at Fcst   05:00
 parallel netcdf      Write Time is   27.19671 at Fcst   06:00
 parallel netcdf      Write Time is   20.22734 at Fcst   06:00
 total            Write Time is   99.69328 at Fcst   06:00

**Comment by jswhit** [comment]
To recap, netcdf chunksizes (for parallel and serial compressed IO) can be set at runtime by specifying parameters ichunk2d,jchunk2d,ichunk3d,jchunk3d,kchunk3d in model_configure.  The default values (if these parameters are not given) are the MPI decomposition size.  If the parameters are set to a negative value, then the netcdf C library will choose the chunksize.  If compression is turned off, the netcdf C library chooses the chunksize (the chunking parameters are not used).

My tests on hera show that the default values produced the fastest parallel IO throughput.  However, this might not be true on other platforms (and Jun's tests suggest it may not be true on WCOSS)  

Note that for parallel IO, the chunksize that produces the fastest write speed may not be optimal for read speed (depending on the access pattern).  For example, my tests indicate that setting the chunksize for 3d variables to imo,jmo,1,1 greatly speeds up reads for 2d slices (a very common access pattern), but slows down the writes by 25% or so on hera.

In general, the write speeds for parallel IO with compression appear to be quite sensitive to chunksize (much less so for serial IO).

The [nccopy](https://www.unidata.ucar.edu/software/netcdf/workshops/2011/utilities/Nccopy.html) utility can be used to change the chunking in an existing netcdf file.

**Comment by jswhit** [comment]
To compile on hera, I made the following change to `modulefiles/hera.intel/fv3`:

```
[ufs-weather-model-parnc]$ git diff modulefiles/hera.intel/fv3
diff --git a/modulefiles/hera.intel/fv3 b/modulefiles/hera.intel/fv3
index a369558..362dea3 100644
--- a/modulefiles/hera.intel/fv3
+++ b/modulefiles/hera.intel/fv3
@@ -50,6 +50,8 @@ module load post/8.0.1
 ##
 module use -a /scratch1/NCEPDEV/nems/emc.nemspara/soft/modulefiles
 module load esmf/8.0.0
+module use -a /scratch2/BMC/gsienkf/whitaker/modulefiles/intel
+module load netcdf-parallel/4.7.3

 ##
 ## load cmake
```

This, and corresponding changes to the wcoss modulefiles, plus the addition of PARALLEL_NETCDF to the cmake files, need to be included in a companion pull request to https://github.com/ufs-community/ufs-weather-model.

**Comment by DusanJovic-NOAA** [comment]
The netcdf-c branch named "ejh_parallel_zlib" described in this comment:

https://github.com/NOAA-EMC/fv3atm/issues/23#issuecomment-572725886

does not exist anymore. Which version of netcdf-c should we use?

**Comment by edwardhartnett** [comment]
Use the netcdf-c master branch. All my changes have been merged to master.

**Comment by DusanJovic-NOAA** [comment]
https://github.com/NOAA-GSD/netcdf-c
or
https://github.com/Unidata/netcdf-c

I see NOAA-GSD:master is 6 commits behind Unidata:master

**Comment by edwardhartnett** [comment]
Use the Unidata master. THe NOAA-GSD one is my fork that I use for working on the Unidata one.

**Comment by junwang-noaa** [comment]
Ed,

just to confirm, I am building netcdf-c from unidata on hera, the revision is 2a34eb2a.

.../nems/emc.nemspara/soft/netcdf_parallel/netcdf-c> git clone https://github.com/Unidata/netcdf-c
.../nems/emc.nemspara/soft/netcdf_parallel/netcdf-c> git branch
* master
[ =0 03:35:41 10034 emc.nemspara@hfe05 ]
.../nems/emc.nemspara/soft/netcdf_parallel/netcdf-c> git log |more

commit 2a34eb2ac5996dc23339bdb72918eb5503393d77
Merge: 2e7234ff 0f5bdafe
Author: Ward Fisher <WardF@users.noreply.github.com>
Date:   Mon Jan 27 17:44:48 2020 -0700

    Merge pull request #1603 from NOAA-GSD/ejh_release_notes

    updated RELEASE_NOTES to include results of recent PR merges

**Comment by junwang-noaa** [comment]
I have problem to build the hdf5/1.10.6 on cray, I have

module load PrgEnv-intel
module rm intel
module rm NetCDF-intel-sandybridge
module load intel/16.3.210

export FC=ftn
export CC=cc
export CXX=CC
export LD=ftn
export LDFLAGS=-L${parlibpath}/lib
export CPPFLAGS=-I${parlibpath}/include
export LD_LIBRARY_PATH=${LD_LIBRARY_PATH}:/gpfs/hps3/emc/nems/noscrub/emc.nemspara/soft/netcdf_parallel/lib

but I got error:
configure:4721: ./conftest

Please verify that both the operating system and the processor support Intel(R) MOVBE, FMA, BMI, LZCNT and AVX2 instructions.

configure:4725: $? = 1
configure:4732: error: in `/gpfs/hps3/emc/nems/noscrub/emc.nemspara/soft/netcdf_parallel/hdf5-1.10.6':
configure:4734: error: cannot run C compiled programs.

Does anybody have any idea?

**Comment by edwardhartnett** [comment]
@junwang-noaa you do have the correct netcdf.

However, I don't know what the HDF5 build problem is. Have you asked HDF5 support?

**Comment by jswhit** [comment]
@junwang-noaa - have a look at config.log file in the build directory after configure fails to find more information as to the cause of failure. Search for "cannot" and examine the lines preceding it.

**Comment by edwardhartnett** [comment]
@jswhit BTW I'm still working on szip and netcdf-fortran. There are some complexities but I hope to have a working branch for you soon...

**Comment by junwang-noaa** [comment]
A detailed configuration testing on wcoss dell is at:

https://docs.google.com/document/d/18vqajgOv3flbS35eNPMnYpFyNiprkJpHkVZpP--dx5o/edit

**Comment by junwang-noaa** [comment]
@jswhit2 The error message I  listed above is from the config.log. Not sure why  the executable compiled from c program can't run on cray.

**Comment by junwang-noaa** [comment]
@edwardhartnett Can you create a tag from the master?

**Comment by edwardhartnett** [comment]
@jswhit2 there is now a branch ejh_szip on netcdf-fortran which has the szip code in the fortran APIs.

@junwang-noaa I cannot create a tag. I am not part of Unidata any longer. ;-) However, I believe they will be doing a release in the next few weeks. (No guarantee though.)

**Comment by junwang-noaa** [comment]
Code is committed. I will open a new issue for installing the hdf/netcdf lib on cray.

**Comment by edwardhartnett** [comment]
@jswhit2 the szip changes have been merged into master branch on Unidata's netcdf-fortran project. So just grab that, and build it against the netcdf-c master you have already built.

Then you can call nf90_def_var_szip() on a variable. Make sure you also turn off deflate. You can't use both deflate and szip.
